### PR TITLE
Temporarily remove query params from viewer iframe on navigate

### LINKED ISSUE #25179: Safari 13: Viewer query params cause document.referrer downgrade
Avoids `Document.referrer` downgrade due to harmless viewer query params like `amp_js_v` and `usqp`.

/cc @ampproject/wg-analytics

### PR DESCRIPTION
**Pull request by choumx**
Closes #25179.

This prevents document.referrer downgrade to eTLD+1, but valid query params will still be lost e.g. `foo.example/article?id=12345`.

- Removes query params from viewer iframe before navigation (Safari 13+, cache origin only).
- Restores query params async (opening new tab) or on "pageshow" (top navigation).

Tested on Safari 13 desktop with UA emulation.

### CODE REVIEW

File: src/service/navigation.js [hunk 1]
```diff
@@ -529,24 +530,72 @@ export class Navigation {
...
+          win.history.replaceState(null, '', original);
+        };
+
+        // For blank_, restore query params after the new page opens.
+        if (target === '_blank') {
+          win.setTimeout(restoreQuery, 0);
+        } else {
+          // For _top etc., wait until page is restored from page cache (bfcache).
+          // https://webkit.org/blog/516/webkit-page-cache-ii-the-unload-event/
+          win.addEventListener('pageshow', function onPageShow(e) {
```
Comment (jridgewell) [inline src/service/navigation.js]: Does this prevent bfcache like unload does?

File: src/service/navigation.js [hunk 2]
```diff
@@ -529,24 +530,72 @@ export class Navigation {
...
+        win.history.replaceState(null, '', noQuery);
+
+        const restoreQuery = () => {
+          dev().info(TAG, 'Restored URL with query string:', original);
+          win.history.replaceState(null, '', original);
+        };
+
+        // For blank_, restore query params after the new page opens.
+        if (target === '_blank') {
+          win.setTimeout(restoreQuery, 0);
```
Comment (jridgewell) [inline src/service/navigation.js]: `timer.delay`?

File: src/service/navigation.js [hunk 3]
```diff
@@ -529,24 +530,72 @@ export class Navigation {
...
+          'Removing query string before navigation:',
+          location.search
+        );
+        const original = location.href;
+        const noQuery = `${location.origin}${location.pathname}${location.hash}`;
+        win.history.replaceState(null, '', noQuery);
+
+        const restoreQuery = () => {
+          dev().info(TAG, 'Restored URL with query string:', original);
+          win.history.replaceState(null, '', original);
```
Comment (jridgewell) [inline src/service/navigation.js]: Depending on the order of timeouts, we may need to check that the URL is still the `noQuery` URL before replacing it.

File: src/service/navigation.js (hunk 2 above)
Comment (choumx) [reply src/service/navigation.js]: Does it buy us anything? Also:

> This uses a micro task for 0

File: src/service/navigation.js (hunk 1 above)
Comment (choumx) [reply src/service/navigation.js]: Nope.

File: src/service/navigation.js (hunk 3 above)
Comment (zhouyx) [reply src/service/navigation.js]: Does this also restore query params that also got appended before navigation. For example the linker param?

Comment (choumx) [reply src/service/navigation.js]: Linker appends a query param to the viewer iframe? I thought it decorates the outgoing link.

In any case, this is the last code that runs before navigation including any URL expansion etc.

Comment (zhouyx) [reply src/service/navigation.js]: oops, nvm I was confused by all the urls....😅 
LGTM

**Comment by zhouyx** [comment]
I noticed you put temporarily to the PR description. Is there any long term solution be considered?

**Comment by choumx** [comment]
This is the long-term solution. :) "Temporarily" is only because we quickly restore the query params after opening a new tab or back navigation.

File: src/service/navigation.js (hunk 2 above)
Comment (jridgewell) [reply src/service/navigation.js]: Explicitly the micro task. Do we need a macro task for this to work?

File: src/service/navigation.js [hunk 4]
```diff
@@ -529,24 +530,82 @@ export class Navigation {
...
+        const restoreQuery = () => {
+          const currentHref = win.location.href;
+          if (currentHref == noQuery) {
+            dev().info(TAG, 'Restored iframe URL with query string:', original);
+            win.history.replaceState(null, '', original);
+          } else {
+            dev().error(
+              TAG,
+              'Unexpected iframe URL change:',
+              currentHref,
```
Comment (jridgewell) [inline src/service/navigation.js]: We can't log these into the Error tracker due to PII.

Comment (choumx) [reply src/service/navigation.js]: Should be safe since it's just the page URL with/without query params.

File: src/service/navigation.js (hunk 2 above)
Comment (choumx) [reply src/service/navigation.js]: Yup, tried it.

**Comment by choumx** [comment]
@jridgewell Need bundle size approval please. :)

**Pull request merged**